processing/src/main/java/io/druid/segment/incremental/OffheapIncrementalIndex.java

```diff
@@ -393,82 +528,143 @@ public OffheapDimDim(String dimension)
      *
      * @see io.druid.segment.incremental.IncrementalIndexStorageAdapter.EntryHolderValueMatcherFactory#makeValueMatcher(String, String)
      */
-    public String get(String str)
+    @Override
+    public String intern(String str)
     {
-      final WeakReference<String> cached = cache.get(str);
-      if (cached != null) {
-        final String value = cached.get();
-        if (value != null) {
-          return value;
+      synchronized (cache) {
+        final WeakReference<String> cached = cache.get(str);
+        if (cached != null) {
+          final String value = cached.get();
+          if (value != null) {
+            return value;
+          }
         }
+        cache.put(str, new WeakReference<String>(str));
+        return str;
       }
-      cache.put(str, new WeakReference(str));
-      return str;
     }
 
+    private static final int ERROR_CLOSED = -2;
+
+    @Override
     public int getId(String value)
     {
+      if (isClosed.get()) {
+        return ERROR_CLOSED; // special error code
+      }
       return falseIds.get(value);
     }
 
+    @Override
     public String getValue(int id)
     {
+      if (id < 0) {
```

Comment:
we typically use Preconditions.checkArgument for those types of checks, which is much nicer to format the error message.